docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
++				rf := NewReferenceField(crd, referenceFieldNames, fieldConfig)
++				crd.SpecFields[referenceFieldNames.Original] = rf
++				crd.Fields[referenceFieldNames.Camel] = rf
++			}
+
+...
+	m.crds = crds
+	return crds, nil
+}
+```
+
+```go
+//field.go
+func NewReferenceField(
+	crd *CRD,
+	fieldNames names.Names,
+	cfg *ackgenconfig.FieldConfig,
+) *Field {
+	gt := "*ackv1alpha1.AWSResourceReference"
+	gtp := "*ackv1alpha1.AWSResourceReference"
+	if cfg.Reference.IsList {
+		gt = "[]" + gt
+		gtp = "[]" + gtp
+	}
+	return &Field{
+		CRD:               crd,
+		Names:             fieldNames,
+		Path:              fieldNames.Original,
+		ShapeRef:          nil,
+		GoType:            gt,
+		GoTypeElem:        "AWSResourceReference",
+		GoTypeWithPkgName: gtp,
+		FieldConfig:       cfg,
+	}
+}
+```
+
+#### Runtime Changes
+
+```go
+// aws_resource_manager.go
+
+type AWSResourceManager interface {
+        ...
++       // DereferenceResource finds if there are any AWSResourceReference fields
++       // present in the AWSResource passed in the parameters and attempts to resolve
++       // those reference fields into target field.
++       // It returns an AWSResource with resolved references, boolean representing
++       // whether any reference fields were present and an error if the passed
++       // AWSResource's reference fields cannot be resolved
++       DereferenceResource(context.Context, client.Reader, AWSResource) (AWSResource, bool, error)
+}
+```
+
+```go
+// reconciler.go
+
+// reconcile either cleans up a deleted resource or ensures that the supplied
+// AWSResource's backing API resource matches the supplied desired state.
+//
+// It returns a copy of the resource that represents the latest observed state.
+func (r *resourceReconciler) reconcile(
+ctx context.Context,
+rm acktypes.AWSResourceManager,
+res acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+    if res.IsBeingDeleted() {
++       // Resolve references before deleting the resource.
++       // Ignore any errors while resolving the references
++    	  res, _, _ = rm.DereferenceResource(ctx, r.apiReader, res)
+        return r.deleteResource(ctx, rm, res)
+    }
+    return r.Sync(ctx, rm, res)
+}
+
+// Sync ensures that the supplied AWSResource's backing API resource
+// matches the supplied desired state.
+//
+// It returns a copy of the resource that represents the latest observed state.
+func (r *resourceReconciler) Sync(
+	ctx context.Context,
+	rm acktypes.AWSResourceManager,
+	desired acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("r.Sync")
+	defer exit(err)
+
+	var latest acktypes.AWSResource // the newly created or mutated resource
+
+	isAdopted := IsAdopted(desired)
+	rlog.WithValues("is_adopted", isAdopted)
++	rlog.Enter("rm.DereferenceResource")
++	dereferencedDesired, containsResourceRef, err := rm.DereferenceResource(ctx, r.apiReader, desired)
```

Comment:
Pretty sure you don't mean `DereferenceResource` here, but rather `ResolveReferences` or something like that?